### PR TITLE
Code Generation Post

### PR DESCRIPTION
**Pull request by rambleraptor**
Adding a post to talk about code generation

### CODE REVIEW

File: content/posts/codegen.md [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+---
+
+# Generating Code
+
+At my day job at Google, I spend a lot of time generating code. Coming from
+outside of Google, this wasn't something I had ever thought about doing.
+Computers writing code for other computers to run has a very Skynet feel to it
+and sounded too complicated to ever be worth the effort.
+Yet, this practice is so commonplace that there's [built-in
+support](https://docs.bazel.build/versions/main/be/general.html#genrule) for it
```
Comment (eashwar) [inline content/posts/codegen.md:15]: CSS nit: you should use the `a:hover` selector so that the UX for links is better. I almost didn't think that this was a hyperlink.

File: content/posts/codegen.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+in Google's build system.
+
+Quick preface: I'm talking about writing a computer program that outputs
+some kind of file that's funny designed to be read in by a compiler or
```
Comment (fishythefish) [inline content/posts/codegen.md:19]:
```suggestion
some kind of file that's designed to be read in by a compiler or
```

File: content/posts/codegen.md [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+interpreter. I'm not talking about writing data files or schema files. It feels
+natural for computers to be writing those, since they broadly match up with the
+native data types found in programming languages.
```
Comment (eashwar) [inline content/posts/codegen.md:22]: I think the grammar here is a little confusing. suggested rewrite for clarity (you don't have to take the way i rewrote it, you can do a different way too):
```suggestion
Quick preface: I'm talking about writing a computer program that outputs
some kind of funny file that's designed to be read by a compiler or
interpreter (not a human). I'm not talking about writing data files or schema files, which 
are pretty natural for computers to be writing, since they broadly match up with the
native data types found in programming languages.
```

File: content/posts/codegen.md [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+If you haven't seen GCP lately, there's an unbelievable amount of resource types
+(in the order of hundreds!). Everytime a user wishes to work with a resource of
+a certain type in one of those tools, there has to be code in that cool that
+handles that particular resource type.
+
+Writing support for each of these resource types looks pretty
+similar. They all follow the same rough logic:
+
+- A user needs a cloud resource to exist in a certain manner
+- The tool calls that cloud resource's GET API to see if the resource exists.
```
Comment (fishythefish) [inline content/posts/codegen.md:43]: GET/Create/update all have different capitalization, better to be consistent

File: content/posts/codegen.md [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+Google Cloud's different resource types for various declarative tooling like
+[Terraform](https://github.com/hashicorp/terraform-provider-google) or
+[Ansible](https://github.com/ansible-collections/google.cloud). We've even
+open-sourced one of our [code
+generators](https://github.com/GoogleCloudPlatform/magic-modules) if you'd like
+to take a look!
+
+If you haven't seen GCP lately, there's an unbelievable amount of resource types
+(in the order of hundreds!). Everytime a user wishes to work with a resource of
+a certain type in one of those tools, there has to be code in that cool that
```
Comment (eashwar) [inline content/posts/codegen.md:36]:
```suggestion
a certain type in one of those tools, there has to be code in that tool that
```

File: content/posts/codegen.md [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+Doing this once or twice is interesting! Doing it a dozen or so times feels
+monotonous, especially once you realize that you made the same bug in all dozen
+resource types and have to make the same fix twelve times. Now, imagine doing
+it a hundred times.
+
+Code generation lets us create a source-of-truth for all this boilerplate.
+
+## How to Generate Code
+
+There's several different approaches to code generation.
```
Comment (fishythefish) [inline content/posts/codegen.md:73]:
```suggestion
There are several different approaches to code generation.
```

File: content/posts/codegen.md [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+- **Compiler Magic:** You could programmatically build out the proper Abstract
+  Syntax Tree and then rely on the compiler to either output the AST as code or
+  just run the AST directly. I suspect that most people handle code generation
+  with some amount of this. That's not actually the method I'm going to be
+  talking about. While you don't have to have a compiler background to do this,
+  it does require knowledge of your language's internals. Testing will be more
+  straightforward of any of these.
```
Comment (fishythefish) [inline content/posts/codegen.md:81]: I don't understand the last sentence.

File: content/posts/codegen.md [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+We primarly use templates for ease of iteration. We can make most changes to the
+templates easily and be reasonably certain that the resulting code will appear
+as we intend. Refactoring the generated code can be done at the template level
+and doesn't require large generator rewrites. We can also programatically
+generate anything we want. We can use the same sets of primatives and
+infrastructure to create Markdown documentation, Go code, Python code,
+protobufs, YAML files and anything in between. 
+
+The downside to this approach is that the templates can grow unwiedy. As we
```
Comment (fishythefish) [inline content/posts/codegen.md:111]:
```suggestion
The downside to this approach is that the templates can grow unwieldy. As we
```

File: content/posts/codegen.md [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+fact that our team needs to write out all of those boilerplate methods...and
+that's before the fact that we realize that the boilerplate isn't just
+copy-paste.
+
+Doing this once or twice is interesting! Doing it a dozen or so times feels
+monotonous, especially once you realize that you made the same bug in all dozen
+resource types and have to make the same fix twelve times. Now, imagine doing
+it a hundred times.
+
+Code generation lets us create a source-of-truth for all this boilerplate.
```
Comment (eashwar) [inline content/posts/codegen.md:69]: Really well written section overall -- it almost feels like you're telling a story, the way you're narratively walking the reader through the motivation here.

File: content/posts/codegen.md [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+
+My team at Google is responsible for building open source support for all of
+Google Cloud's different resource types for various declarative tooling like
+[Terraform](https://github.com/hashicorp/terraform-provider-google) or
+[Ansible](https://github.com/ansible-collections/google.cloud). We've even
+open-sourced one of our [code
+generators](https://github.com/GoogleCloudPlatform/magic-modules) if you'd like
+to take a look!
+
+If you haven't seen GCP lately, there's an unbelievable amount of resource types
```
Comment (fishythefish) [inline content/posts/codegen.md:34]:
```suggestion
If you haven't seen GCP lately, there's an unbelievable number of resource types
```

File: content/posts/codegen.md [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+(in the order of hundreds!). Everytime a user wishes to work with a resource of
```
Comment (fishythefish) [inline content/posts/codegen.md:35]:
```suggestion
(on the order of hundreds!). Every time a user wishes to work with a resource of
```

File: content/posts/codegen.md (hunk 5 above)
Comment (fishythefish) [inline content/posts/codegen.md:36]:
```suggestion
a certain type in one of those tools, there has to be code in that tool that
```

File: content/posts/codegen.md [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+a certain type in one of those tools, there has to be code in that cool that
+handles that particular resource type.
+
+Writing support for each of these resource types looks pretty
+similar. They all follow the same rough logic:
+
+- A user needs a cloud resource to exist in a certain manner
```
Comment (fishythefish) [inline content/posts/codegen.md:42]:
```suggestion
- A user needs a cloud resource to exist in a certain manner.
```

File: content/posts/codegen.md [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+  defeats the purpose of automation.
+
+While many people will take one of these approaches (especially the first one),
+we've chosen not to. We actually take a third approach: templating.
+
+## Templated Approach
+
+The approach we take is essentially Mad Libs.
+
+<img src="/codegen.png" class="img-fluid hero-image" alt="Codegen example">
```
Comment (eashwar) [inline content/posts/codegen.md:96]: Nit: alt text here could be more descriptive for accessibility purposes.

File: content/posts/codegen.md [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+The downside to this approach is that the templates can grow unwiedy. As we
+discover more places where customizations need to occur, those customizations
+become embedded into our templates. IDEs have poor support for understanding
+template syntax, leaving it up to the team to carefully parse our templates and
+understand what's happening.
+
+## Abstractions: When to Generate?
+
+Typically, when writing code, we think about abstractions. We start out with a
+list of instructions. We can create functions, which abstract away those list
```
Comment (fishythefish) [inline content/posts/codegen.md:120]:
```suggestion
list of instructions. We can create functions, which abstract away those lists
```

But also consider just "those instructions."

File: content/posts/codegen.md [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+of instructions. Then, we can create classes, which abstract away both the
+functions and the underlying data used by the functions. Code generation acts
+as a layer on top of all of these.
+
+Because of this, we're constantly asking ourselves if we're using the right
+abstraction. Human-written code is easy to test. Generated code is easy to
+test, but the generator itself is hard to test. This means that we're always on
+the hunt for more edge cases in generated code, so we can improve the generator
+better.
```
Comment (fishythefish) [inline content/posts/codegen.md:129]:
```suggestion
the hunt for more edge cases in generated code, so we can improve the generator.
```

File: content/posts/codegen.md [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+  insight into what a ML model is doing or how to debug it. Your training data
+  is probably going to look a lot like your final product, which sort of
+  defeats the purpose of automation.
+
+While many people will take one of these approaches (especially the first one),
+we've chosen not to. We actually take a third approach: templating.
+
+## Templated Approach
+
+The approach we take is essentially Mad Libs.
```
Comment (eashwar) [inline content/posts/codegen.md:94]: idea: I think it may be useful to start this section off by talking about templating being used with strings in languages, like C# string interpolation or C++ `std::format`, as these are things many people may have experience with. you could then motivate that you're applying the same principals but at the scale of entire files and more.

File: content/posts/codegen.md [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+Because of this, we're constantly asking ourselves if we're using the right
+abstraction. Human-written code is easy to test. Generated code is easy to
+test, but the generator itself is hard to test. This means that we're always on
+the hunt for more edge cases in generated code, so we can improve the generator
+better.
+
+I suspect that this is why using "compiler magic" is so much more tenable for
+teams. It's easier to write unit tests when you have handfuls of functions that
+all alter an AST. That's how developers learn to write code! Write a function
+and then write a text for that function. A template approach means that most of
```
Comment (fishythefish) [inline content/posts/codegen.md:134]:
```suggestion
and then write a test for that function. A template approach means that most of
```

File: content/posts/codegen.md [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+the true "logic" of your program is embedded inside a templating library that
+you have no insight into. You can test the end result, but you can't easily
+inject tests into anything in the middle.
+
+Our team does rewrites where we take a large amount of template, write a
+function that replicates that template at runtime, and then replace the
+template with a single line of template code. We can more easily test our
```
Comment (fishythefish) [inline content/posts/codegen.md:141]: This sentence is kinda "yo dawg, I heard you like templates". Consider restructuring, especially to distinguish the different uses of "template".

File: content/posts/codegen.md [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+handwritten function. At that point, we just have to make sure that the
```
Comment (fishythefish) [inline content/posts/codegen.md:142]:
```suggestion
handwritten function than the original template. At that point, we just have to make sure that the
```

File: content/posts/codegen.md [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+template places the correct values into the function.
```
Comment (fishythefish) [inline content/posts/codegen.md:143]: "template" here is also a little confusing because of the overuse above.

Comment (eashwar) [inline content/posts/codegen.md:143]: Love this section. I think it doesn't feel like a conclusion though, it feels like it abruptly ends here. Might suggest writing some sort of conclusion or something!

File: content/posts/codegen.md [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+and doesn't require large generator rewrites. We can also programatically
+generate anything we want. We can use the same sets of primatives and
+infrastructure to create Markdown documentation, Go code, Python code,
+protobufs, YAML files and anything in between. 
+
+The downside to this approach is that the templates can grow unwiedy. As we
+discover more places where customizations need to occur, those customizations
+become embedded into our templates. IDEs have poor support for understanding
+template syntax, leaving it up to the team to carefully parse our templates and
+understand what's happening.
```
Comment (eashwar) [inline content/posts/codegen.md:115]: glad you discuss the cons to this approach too!

**Review by eashwar — COMMENTED** [review]
great article overall! had a few comments here and there

**Review by fishythefish — COMMENTED** [review]
Mostly some quick nitpicks, but I haven't included thoughts on the overall structure or anything. The main thing I noticed was that the post ends rather abruptly.
